File: src/main/kotlin/de/rewe/kafkatogcs/schema/conversion/JsonToAvroSchemaConverter.java
package de.rewe.kafkatogcs.schema.conversion;

import static java.util.stream.Collectors.toList;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import de.rewe.kafkatogcs.schema.conversion.commons.IteratorsUtils;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;
import javax.annotation.Nullable;
import org.apache.avro.LogicalTypes;
import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;
import org.apache.flink.shaded.guava30.com.google.common.base.Preconditions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The main function of this class is to convert a JsonSchema to Avro schema. It can also
 * standardize schema names, and keep track of a mapping from the original names to the standardized
 * ones, which is needed for unit tests. <br/>
 */
public class JsonToAvroSchemaConverter {

  private static final String TYPE = "type";
  private static final Schema NULL_SCHEMA = Schema.create(Schema.Type.NULL);
  private static final Schema STRING_SCHEMA = Schema.create(Schema.Type.STRING);
  private static final Logger LOGGER = LoggerFactory.getLogger(JsonToAvroSchemaConverter.class);

  static List<JsonSchemaType> getNonNullTypes(
      final String fieldName,
      final JsonNode fieldDefinition) {
    return getTypes(fieldName, fieldDefinition).stream()
        .filter(type -> type != JsonSchemaType.NULL).collect(toList());
  }

  /**
   * When no type is specified, it will default to string.
   */
  static List<JsonSchemaType> getTypes(final String fieldName, final JsonNode fieldDefinition) {
    final Optional<JsonNode> combinedRestriction = getCombinedRestriction(fieldDefinition);
    if (combinedRestriction.isPresent()) {
      return List.of(JsonSchemaType.COMBINED);
    }

    final JsonNode typeProperty = fieldDefinition.get(TYPE);
    if (typeProperty == null || typeProperty.isNull()) {
      LOGGER.warn("Field \"{}\" has no type specification. It will default to string", fieldName);
      return List.of(JsonSchemaType.STRING);
    }

    if (typeProperty.isArray()) {
      return IteratorsUtils.toList(typeProperty.elements()).stream()
                           .map(s -> JsonSchemaType.fromJsonSchemaTypeName(s.asText()))
                           .collect(toList());
    }

    if (typeProperty.isTextual()) {
      return List.of(JsonSchemaType.fromJsonSchemaTypeName(typeProperty.asText()));
    }

    LOGGER.warn("Field \"{}\" has unexpected type {}. It will default to string.",
        fieldName,
        typeProperty);
    return List.of(JsonSchemaType.STRING);
  }

  static Optional<JsonNode> getCombinedRestriction(final JsonNode fieldDefinition) {
    if (fieldDefinition.has("anyOf")) {
      return Optional.of(fieldDefinition.get("anyOf"));
    }
    if (fieldDefinition.has("allOf")) {
      return Optional.of(fieldDefinition.get("allOf"));
    }
    if (fieldDefinition.has("oneOf")) {
      return Optional.of(fieldDefinition.get("oneOf"));
    }
    return Optional.empty();
  }

  /**
   * @return Avro schema based on the input {@code jsonSchema}.
   */
  public Schema getAvroSchema(final JsonNode jsonSchema,
                              final String streamName,
                              @Nullable final String namespace) {
    return getAvroSchema(jsonSchema, streamName, namespace, true, true, true);
  }

  /**
   * @param appendExtraProps Add default additional property field to the output Avro schema.
   * @param addStringToLogicalTypes Default logical type field to string.
   * @param isRootNode Whether it is the root field in the input Json schema.
   * @return Avro schema based on the input {@code jsonSchema}.
   */
  public Schema getAvroSchema(final JsonNode jsonSchema,
                              final String fieldName,
                              @Nullable final String fieldNamespace,
                              final boolean appendExtraProps,
                              final boolean addStringToLogicalTypes,
                              final boolean isRootNode) {
    final String stdName = AvroConstants.NAME_TRANSFORMER.getIdentifier(fieldName);
    final String stdNamespace = AvroConstants.NAME_TRANSFORMER.getNamespace(fieldNamespace);
    final SchemaBuilder.RecordBuilder<Schema> builder = SchemaBuilder.record(stdName);
    if (!stdName.equals(fieldName)) {
      LOGGER.warn("Schema name \"{}\" contains illegal character(s) and is standardized to \"{}\"",
          fieldName,
          stdName);
      builder.doc(
          String.format("%s",
              fieldName));
    }
    if (stdNamespace != null) {
      builder.namespace(stdNamespace);
    }

    final Optional<JsonNode> properties = Optional.ofNullable(jsonSchema.get("properties"));
    // object field with no "properties" will be handled by the default additional properties
    // field during object conversion; so it is fine if there is no "properties"
    final List<String> subfieldNames = properties
        .map(p -> IteratorsUtils.toList(p.fieldNames()))
        .orElseGet(List::of);

    final SchemaBuilder.FieldAssembler<Schema> assembler = builder.fields();

    for (final String subfieldName : subfieldNames) {
      final String stdFieldName = AvroConstants.NAME_TRANSFORMER.getIdentifier(subfieldName);
      final JsonNode subfieldDefinition = properties.map(p -> p.get(subfieldName)).orElseGet(
          JsonNodeFactory.instance::missingNode);
      final SchemaBuilder.FieldBuilder<Schema> fieldBuilder = assembler.name(stdFieldName);
      if (!stdFieldName.equals(subfieldName)) {
        LOGGER.warn("Field name \"{}\" contains illegal character(s) and is standardized to \"{}\"",
            subfieldName, stdFieldName);
        fieldBuilder.doc(String.format("%s",
            subfieldName));
      }
      final String subfieldNamespace = isRootNode
          // Omit the namespace for root level fields,
          // because it is directly assigned in the builder above.
          // This may not be the correct choice.
          ? null
          : (stdNamespace == null ? stdName : (stdNamespace + "." + stdName));
      fieldBuilder.type(parseJsonField(
              subfieldName,
              subfieldNamespace,
              subfieldDefinition,
              appendExtraProps,
              addStringToLogicalTypes))
          .withDefault(null);
    }

    return assembler.endRecord();
  }

  /**
   * Generate Avro schema for a single Json field type. For example:
   *
   * <pre>
   * "number" -> ["double"]
   * </pre>
   */
  Schema parseSingleType(final String fieldName,
                         @Nullable final String fieldNamespace,
                         final JsonSchemaType fieldType,
                         final JsonNode fieldDefinition,
                         final boolean appendExtraProps,
                         final boolean addStringToLogicalTypes) {
    Preconditions.checkState(
            fieldType != JsonSchemaType.NULL,
            "Null types should have been filtered out");

    final Schema fieldSchema;
    switch (fieldType) {
      case INTEGER:
      case LONG:
      case NUMBER:
      case BOOLEAN:
        fieldSchema = Schema.create(fieldType.getAvroType());
        break;
      case STRING:
        fieldSchema = getStringSchema(fieldType, fieldDefinition);
        break;
      case COMBINED:
        fieldSchema = getCombinedSchema(
            fieldName,
            fieldNamespace,
            fieldDefinition,
            appendExtraProps,
            addStringToLogicalTypes);
        break;
      case ARRAY:
        fieldSchema = getArraySchema(
            fieldName,
            fieldNamespace,
            fieldDefinition,
            appendExtraProps,
            addStringToLogicalTypes);
        break;
      case OBJECT:
        fieldSchema = getAvroSchema(
            fieldDefinition,
            fieldName,
            fieldNamespace,
            appendExtraProps,
            addStringToLogicalTypes,
            false);
        break;
      default:
        LOGGER.warn(
            "Field \"{}\" has invalid type definition: {}. It will default to string.",
            fieldName,
            fieldDefinition);
        fieldSchema = Schema.createUnion(NULL_SCHEMA, STRING_SCHEMA);
        break;
    }
    return fieldSchema;
  }

  private Schema getArraySchema(
      String fieldName,
      String fieldNamespace,
      JsonNode fieldDefinition,
      boolean appendExtraProps,
      boolean addStringToLogicalTypes) {
    final Schema fieldSchema;
    final JsonNode items = fieldDefinition.get("items");
    if (items == null) {
      LOGGER.warn("Array field \"{}\" does not specify the items type. It will default to an array of strings",
          fieldName);
      fieldSchema = Schema.createArray(Schema.createUnion(NULL_SCHEMA, STRING_SCHEMA));
    } else if (items.isObject()) {
      if (!items.has("type") || items.get("type").isNull()) {
        LOGGER.warn("Array field \"{}\" does not specify the items type. it will default to an array of strings",
            fieldName);
        fieldSchema = Schema.createArray(Schema.createUnion(NULL_SCHEMA, STRING_SCHEMA));
      } else {
        // Objects inside Json array has no names. We name it with the ".items" suffix.
        final String elementFieldName = fieldName + ".items";
        fieldSchema = Schema.createArray(parseJsonField(elementFieldName,
            fieldNamespace, items,
            appendExtraProps,
            addStringToLogicalTypes));
      }
    } else if (items.isArray()) {
      final List<Schema> arrayElementTypes =
          parseJsonTypeUnion(
              fieldName,
              fieldNamespace, (ArrayNode) items,
              appendExtraProps,
              addStringToLogicalTypes);
      arrayElementTypes.add(0, NULL_SCHEMA);
      fieldSchema = Schema.createArray(Schema.createUnion(arrayElementTypes));
    } else {
      LOGGER.warn("Array field \"{}\" has invalid items specification: {}. It will default to an array of strings.",
          fieldName, items);
      fieldSchema = Schema.createArray(Schema.createUnion(NULL_SCHEMA, STRING_SCHEMA));
    }
    return fieldSchema;
  }

  private Schema getCombinedSchema(
      String fieldName,
      String fieldNamespace,
      JsonNode fieldDefinition,
      boolean appendExtraProps,
      boolean addStringToLogicalTypes) {
    final Schema fieldSchema;
    final Optional<JsonNode> combinedRestriction = getCombinedRestriction(fieldDefinition);
    final List<Schema> unionTypes =
        parseJsonTypeUnion(fieldName, fieldNamespace, (ArrayNode) combinedRestriction.orElseGet(
            JsonNodeFactory.instance::arrayNode), appendExtraProps, addStringToLogicalTypes);
    fieldSchema = Schema.createUnion(unionTypes);
    return fieldSchema;
  }

  private static Schema getStringSchema(JsonSchemaType fieldType, JsonNode fieldDefinition) {
    final Schema fieldSchema;
    if (fieldDefinition.has("format")) {
      final String format = fieldDefinition.get("format").asText();
      if ("date-time".equals(format)) {
        fieldSchema = LogicalTypes.timestampMicros()
                                  .addToSchema(Schema.create(Schema.Type.LONG));
      } else if ("date".equals(format)) {
        fieldSchema = LogicalTypes.date()
                                  .addToSchema(Schema.create(Schema.Type.INT));
      } else if ("time".equals(format)) {
        fieldSchema = LogicalTypes.timeMicros()
                                  .addToSchema(Schema.create(Schema.Type.LONG));
      } else {
        fieldSchema = Schema.create(fieldType.getAvroType());
      }
    } else {
      fieldSchema = Schema.create(fieldType.getAvroType());
    }
    return fieldSchema;
  }

  /**
   * Take in a union of Json field definitions, and generate Avro field schema unions. For example:
   *
   * <pre>
   * ["number", { ... }] -> ["double", { ... }]
   * </pre>
   */
  List<Schema> parseJsonTypeUnion(final String fieldName,
                                  @Nullable final String fieldNamespace,
                                  final ArrayNode types,
                                  final boolean appendExtraProps,
                                  final boolean addStringToLogicalTypes) {
    final List<Schema> schemas = IteratorsUtils.toList(types.elements())
        .stream()
        .flatMap(definition -> getNonNullTypes(fieldName, definition).stream().flatMap(type -> {
          final String namespace = fieldNamespace == null
              ? fieldName
              : fieldNamespace + "." + fieldName;
          final Schema singleFieldSchema = parseSingleType(
              fieldName,
              namespace,
              type,
              definition,
              appendExtraProps,
              addStringToLogicalTypes);

          if (singleFieldSchema.isUnion()) {
            return singleFieldSchema.getTypes().stream();
          } else {
            return Stream.of(singleFieldSchema);
          }
        }))
        .distinct()
        .collect(toList());

    return mergeRecordSchemas(fieldName, fieldNamespace, schemas);
  }

  /**
   * If there are multiple object fields, those fields are combined into one Avro record. This is
   * because Avro does not allow specifying a tuple of types (i.e. the first element is type x, the
   * second element is type y, and so on). For example, the following Json field types:
   *
   * <pre>
   * [
   *   {
   *     "type": "object",
   *     "properties": {
   *       "id": { "type": "integer" }
   *     }
   *   },
   *   {
   *     "type": "object",
   *     "properties": {
   *       "id": { "type": "string" }
   *       "message": { "type": "string" }
   *     }
   *   }
   * ]
   * </pre>
   *
   * is converted to this Avro schema:
   *
   * <pre>
   * {
   *   "type": "record",
   *   "fields": [
   *     { "name": "id", "type": ["int", "string"] },
   *     { "name": "message", "type": "string" }
   *   ]
   * }
   * </pre>
   */
  List<Schema> mergeRecordSchemas(final String fieldName,
                                  @Nullable final String fieldNamespace,
                                  final List<Schema> schemas) {
    final LinkedHashMap<String, List<Schema>> recordFieldSchemas = new LinkedHashMap<>();
    final Map<String, List<String>> recordFieldDocs = new HashMap<>();

    final List<Schema> mergedSchemas = schemas.stream()
        // gather record schemas to construct a single record schema later on
        .peek(schema -> {
          if (schema.getType() == Schema.Type.RECORD) {
            for (final Schema.Field field : schema.getFields()) {
              recordFieldSchemas.putIfAbsent(field.name(), new LinkedList<>());
              recordFieldSchemas.get(field.name()).add(field.schema());
              if (field.doc() != null) {
                recordFieldDocs.putIfAbsent(field.name(), new LinkedList<>());
                recordFieldDocs.get(field.name()).add(field.doc());
              }
            }
          }
        })
        // remove record schemas because they will be merged into one
        .filter(schema -> schema.getType() != Schema.Type.RECORD)
        .collect(toList());

    // create one record schema from all the record fields
    if (!recordFieldSchemas.isEmpty()) {
      final SchemaBuilder.RecordBuilder<Schema> builder = SchemaBuilder.record(fieldName);
      if (fieldNamespace != null) {
        builder.namespace(fieldNamespace);
      }

      final SchemaBuilder.FieldAssembler<Schema> assembler = builder.fields();

      for (final Map.Entry<String, List<Schema>> entry : recordFieldSchemas.entrySet()) {
        final String subfieldName = entry.getKey();

        final SchemaBuilder.FieldBuilder<Schema> subfieldBuilder = assembler.name(subfieldName);
        final List<String> subfieldDocs =
            recordFieldDocs.getOrDefault(subfieldName, List.of());
        if (!subfieldDocs.isEmpty()) {
          subfieldBuilder.doc(String.join("; ", subfieldDocs));
        }
        final List<Schema> subfieldSchemas = entry.getValue().stream()
            .flatMap(schema -> schema.getTypes().stream()
                // filter out null and add it later on as the first element
                .filter(s -> !s.equals(NULL_SCHEMA)))
            .distinct()
            .collect(toList());
        // recursively merge schemas of a subfield
        // because they may include multiple record schemas as well
        final List<Schema> mergedSubfieldSchemas =
            mergeRecordSchemas(subfieldName, fieldNamespace, subfieldSchemas);
        mergedSubfieldSchemas.add(0, NULL_SCHEMA);
        subfieldBuilder.type(Schema.createUnion(mergedSubfieldSchemas)).withDefault(null);
      }

      mergedSchemas.add(assembler.endRecord());
    }

    return mergedSchemas;
  }

  /**
   * Take in a Json field definition, and generate a nullable Avro field schema. For example:
   *
   * <pre>
   * {"type": ["number", { ... }]} -> ["null", "double", { ... }]
   * </pre>
   */
  Schema parseJsonField(final String fieldName,
                        @Nullable final String fieldNamespace,
                        final JsonNode fieldDefinition,
                        final boolean appendExtraProps,
                        final boolean addStringToLogicalTypes) {
    // Filter out null types, which will be added back in the end.
    final List<Schema> nonNullFieldTypes = getNonNullTypes(fieldName, fieldDefinition)
        .stream()
        .flatMap(fieldType -> {
          final Schema singleFieldSchema =
              parseSingleType(
                  fieldName,
                  fieldNamespace,
                  fieldType,
                  fieldDefinition,
                  appendExtraProps,
                  addStringToLogicalTypes);
          if (singleFieldSchema.isUnion()) {
            return singleFieldSchema.getTypes().stream();
          } else {
            return Stream.of(singleFieldSchema);
          }
        })
        .distinct()
        .collect(toList());

    if (nonNullFieldTypes.isEmpty()) {
      return Schema.create(Schema.Type.NULL);
    } else {
      // Mark every field as nullable to prevent missing value exceptions from Avro / Parquet.
      if (!nonNullFieldTypes.contains(NULL_SCHEMA)) {
        nonNullFieldTypes.add(0, NULL_SCHEMA);
      }
      // Logical types are converted to a union of
      // logical type itself and string. The purpose is to
      // default the logical type field to a string,
      // if the value of the logical type field is invalid and
      // cannot be properly processed.
      if ((nonNullFieldTypes
          .stream().anyMatch(schema -> schema.getLogicalType() != null)) &&
          (!nonNullFieldTypes.contains(STRING_SCHEMA)) && addStringToLogicalTypes) {
        nonNullFieldTypes.add(STRING_SCHEMA);
      }
      return Schema.createUnion(nonNullFieldTypes);
    }
  }

}

File: src/main/kotlin/de/rewe/kafkatogcs/schema/conversion/JsonToAvroSchemaConverter.kt
package de.rewe.kafkatogcs.schema.conversion

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.JsonNodeFactory
import org.apache.avro.LogicalTypes
import org.apache.avro.Schema
import org.apache.avro.SchemaBuilder
import org.apache.flink.shaded.guava30.com.google.common.base.Preconditions
import org.slf4j.LoggerFactory
import java.util.Optional
import kotlin.collections.HashMap
import kotlin.collections.LinkedHashMap

/**
 * The main function of this class is to convert a JsonSchema to Avro schema. It can also
 * standardize schema names, and keep track of a mapping from the original names to the standardized
 * ones, which is needed for unit tests. <br></br>
 */
class JsonToAvroSchemaConverter {
    /**
     * @return Avro schema based on the input `jsonSchema`.
     */
    fun getAvroSchema(
        jsonSchema: JsonNode,
        streamName: String,
        namespace: String?
    ): Schema {
        return getAvroSchema(
            jsonSchema = jsonSchema,
            fieldName = streamName,
            fieldNamespace = namespace,
            appendExtraProps = true,
            addStringToLogicalTypes = true,
            isRootNode = true
        )
    }

    /**
     * @param appendExtraProps Add default additional property field to the output Avro schema.
     * @param addStringToLogicalTypes Default logical type field to string.
     * @param isRootNode Whether it is the root field in the input Json schema.
     * @return Avro schema based on the input `jsonSchema`.
     */
    fun getAvroSchema(
        jsonSchema: JsonNode,
        fieldName: String,
        fieldNamespace: String?,
        appendExtraProps: Boolean,
        addStringToLogicalTypes: Boolean,
        isRootNode: Boolean
    ): Schema {
        val stdName = AvroConstants.NAME_TRANSFORMER.getIdentifier(fieldName)
        val stdNamespace = fieldNamespace?.let { AvroConstants.NAME_TRANSFORMER.getNamespace(it) }
        val builder = SchemaBuilder.record(stdName)
        if (stdName != fieldName) {
            LOGGER.warn(
                "Schema name \"{}\" contains illegal character(s) and is standardized to \"{}\"",
                fieldName,
                stdName
            )
            builder.doc(
                String.format(
                    "%s",
                    fieldName
                )
            )
        }
        if (stdNamespace != null) {
            builder.namespace(stdNamespace)
        }
        val properties = Optional.ofNullable(
            jsonSchema["properties"]
        )
        // object field with no "properties" will be handled by the default additional properties
        // field during object conversion; so it is fine if there is no "properties"
        val subfieldNames = properties
            .map { p: JsonNode -> p.fieldNames().asSequence().toList() }
            .orElseGet { listOf() }
        val assembler = builder.fields()
        for (subfieldName in subfieldNames) {
            val stdFieldName = AvroConstants.NAME_TRANSFORMER.getIdentifier(subfieldName)
            val subfieldDefinition = properties.map { p: JsonNode -> p[subfieldName] }
                .orElseGet { JsonNodeFactory.instance.missingNode() }
            val fieldBuilder = assembler.name(stdFieldName)
                LOGGER.warn(
                    "Field name \"{}\" contains illegal character(s) and is standardized to \"{}\"",
                    subfieldName, stdFieldName
                )
                fieldBuilder.doc(
                    String.format(
                        "%s",
                        subfieldName
                    )
                )
            val subfieldNamespace = if (isRootNode // Omit the namespace for root level fields,
            // because it is directly assigned in the builder above.
            // This may not be the correct choice.
            ) null else if (stdNamespace == null) stdName else "$stdNamespace.$stdName"
            fieldBuilder.type(
                parseJsonField(
                    subfieldName,
                    subfieldNamespace,
                    subfieldDefinition,
                    appendExtraProps,
                    addStringToLogicalTypes
                )
            )
                .withDefault(null)
        }
        return assembler.endRecord()
    }

    /**
     * Generate Avro schema for a single Json field type. For example:
     *
     * <pre>
     * "number" -> ["double"]
    </pre> *
     */
    private fun parseSingleType(
        fieldName: String,
        fieldNamespace: String?,
        fieldType: JsonSchemaType,
        fieldDefinition: JsonNode,
        appendExtraProps: Boolean,
        addStringToLogicalTypes: Boolean
    ): Schema {
        Preconditions.checkState(
            fieldType != JsonSchemaType.NULL,
            "Null types should have been filtered out"
        )
        val fieldSchema: Schema = when (fieldType) {
            JsonSchemaType.INTEGER, JsonSchemaType.LONG, JsonSchemaType.NUMBER, JsonSchemaType.BOOLEAN -> Schema.create(
                fieldType.avroType
            )

            JsonSchemaType.STRING -> getStringSchema(
                fieldType,
                fieldDefinition
            )

            JsonSchemaType.COMBINED -> getCombinedSchema(
                fieldName,
                fieldNamespace,
                fieldDefinition,
                appendExtraProps,
                addStringToLogicalTypes
            )

            JsonSchemaType.ARRAY -> getArraySchema(
                fieldName,
                fieldNamespace,
                fieldDefinition,
                appendExtraProps,
                addStringToLogicalTypes
            )

            JsonSchemaType.OBJECT -> getAvroSchema(
                fieldDefinition,
                fieldName,
                fieldNamespace,
                appendExtraProps,
                addStringToLogicalTypes,
                false
            )

            else -> {
                LOGGER.warn(
                    "Field \"{}\" has invalid type definition: {}. It will default to string.",
                    fieldName,
                    fieldDefinition
                )
                Schema.createUnion(
                    NULL_SCHEMA,
                    STRING_SCHEMA
                )
            }
        }
        return fieldSchema
    }

    private fun getArraySchema(
        fieldName: String,
        fieldNamespace: String?,
        fieldDefinition: JsonNode,
        appendExtraProps: Boolean,
        addStringToLogicalTypes: Boolean
    ): Schema {
        val fieldSchema: Schema
        val items = fieldDefinition["items"]
        fieldSchema = if (items == null) {
            LOGGER.warn(
                "Array field \"{}\" does not specify the items type. It will default to an array of strings",
                fieldName
            )
            Schema.createArray(
                Schema.createUnion(
                    NULL_SCHEMA,
                    STRING_SCHEMA
                )
            )
        } else if (items.isObject) {
            if (!items.has("type") || items["type"].isNull) {
                LOGGER.warn(
                    "Array field \"{}\" does not specify the items type. it will default to an array of strings",
                    fieldName
                )
                Schema.createArray(
                    Schema.createUnion(
                        NULL_SCHEMA,
                        STRING_SCHEMA
                    )
                )
            } else {
                // Objects inside Json array has no names. We name it with the ".items" suffix.
                val elementFieldName = "$fieldName.items"
                Schema.createArray(
                    parseJsonField(
                        elementFieldName,
                        fieldNamespace, items,
                        appendExtraProps,
                        addStringToLogicalTypes
                    )
                )
            }
        } else if (items.isArray) {
            val arrayElementTypes = parseJsonTypeUnion(
                fieldName,
                fieldNamespace, items as ArrayNode,
                appendExtraProps,
                addStringToLogicalTypes
            )
            arrayElementTypes.add(0, NULL_SCHEMA)
            Schema.createArray(Schema.createUnion(arrayElementTypes))
        } else {
            LOGGER.warn(
                "Array field \"{}\" has invalid items specification: {}. It will default to an array of strings.",
                fieldName, items
            )
            Schema.createArray(
                Schema.createUnion(
                    NULL_SCHEMA,
                    STRING_SCHEMA
                )
            )
        }
        return fieldSchema
    }

    private fun getCombinedSchema(
        fieldName: String,
        fieldNamespace: String?,
        fieldDefinition: JsonNode,
        appendExtraProps: Boolean,
        addStringToLogicalTypes: Boolean
    ): Schema {
        val fieldSchema: Schema
        val combinedRestriction = getCombinedRestriction(fieldDefinition)
        val unionTypes = parseJsonTypeUnion(
            fieldName,
            fieldNamespace,
            combinedRestriction.orElseGet { JsonNodeFactory.instance.arrayNode() } as ArrayNode,
            appendExtraProps,
            addStringToLogicalTypes)
        fieldSchema = Schema.createUnion(unionTypes)
        return fieldSchema
    }

    /**
     * Take in a union of Json field definitions, and generate Avro field schema unions. For example:
     *
     * <pre>
     * ["number", { ... }] -> ["double", { ... }]
    </pre> *
     */
    private fun parseJsonTypeUnion(
        fieldName: String,
        fieldNamespace: String?,
        types: ArrayNode,
        appendExtraProps: Boolean,
        addStringToLogicalTypes: Boolean
    ): MutableList<Schema> {
        val schemas = types.elements().asSequence()
            .flatMap { definition: JsonNode ->
                getNonNullTypes(fieldName, definition).flatMap inner@{ type: JsonSchemaType ->
                    val namespace =
                        if (fieldNamespace == null) fieldName else "$fieldNamespace.$fieldName"
                    val singleFieldSchema = parseSingleType(
                        fieldName,
                        namespace,
                        type,
                        definition,
                        appendExtraProps,
                        addStringToLogicalTypes
                    )
                    return@inner if (singleFieldSchema.isUnion) {
                        singleFieldSchema.types.asSequence()
                    } else {
                        sequenceOf(singleFieldSchema)
                    }
                }
            }
            .distinct()
            .toList()
        return mergeRecordSchemas(fieldName, fieldNamespace, schemas)
    }

    /**
     * If there are multiple object fields, those fields are combined into one Avro record. This is
     * because Avro does not allow specifying a tuple of types (i.e. the first element is type x, the
     * second element is type y, and so on). For example, the following Json field types:
     *
     * <pre>
     * [
     * {
     * "type": "object",
     * "properties": {
     * "id": { "type": "integer" }
     * }
     * },
     * {
     * "type": "object",
     * "properties": {
     * "id": { "type": "string" }
     * "message": { "type": "string" }
     * }
     * }
     * ]
    </pre> *
     *
     * is converted to this Avro schema:
     *
     * <pre>
     * {
     * "type": "record",
     * "fields": [
     * { "name": "id", "type": ["int", "string"] },
     * { "name": "message", "type": "string" }
     * ]
     * }
    </pre> *
     */
    private fun mergeRecordSchemas(
        fieldName: String?,
        fieldNamespace: String?,
        schemas: List<Schema>
    ): MutableList<Schema> {
        val recordFieldSchemas = LinkedHashMap<String, MutableList<Schema>>()
        val recordFieldDocs: MutableMap<String, MutableList<String>> = HashMap()
        val mergedSchemas =
            schemas.toList() // gather record schemas to construct a single record schema later on
                .map { schema: Schema ->
                    if (schema.type == Schema.Type.RECORD) {
                        for (field in schema.fields) {
                            recordFieldSchemas.putIfAbsent(field.name(), mutableListOf())
                            recordFieldSchemas[field.name()]!!.add(field.schema())
                            if (field.doc() != null) {
                                recordFieldDocs.putIfAbsent(field.name(), mutableListOf())
                                recordFieldDocs[field.name()]!!.add(field.doc())
                            }
                        }
                    }
                    schema
                } // remove record schemas because they will be merged into one
                .filter { schema: Schema -> schema.type != Schema.Type.RECORD }
                .toMutableList()

        // create one record schema from all the record fields
        if (recordFieldSchemas.isNotEmpty()) {
            val builder = SchemaBuilder.record(fieldName)
            if (fieldNamespace != null) {
                builder.namespace(fieldNamespace)
            }
            val assembler = builder.fields()
            for ((subfieldName, value) in recordFieldSchemas) {
                val subfieldBuilder = assembler.name(subfieldName)
                val subfieldDocs = recordFieldDocs.getOrDefault(subfieldName, listOf())
                if (subfieldDocs.isNotEmpty()) {
                    subfieldBuilder.doc(java.lang.String.join("; ", subfieldDocs))
                }
                val subfieldSchemas = value.toList()
                    .flatMap { schema: Schema ->
                        schema.types.toList() // filter out null and add it later on as the first element
                            .filter { s: Schema -> s != NULL_SCHEMA }
                    }
                    .distinct()
                // recursively merge schemas of a subfield
                // because they may include multiple record schemas as well
                val mergedSubfieldSchemas =
                    mergeRecordSchemas(subfieldName, fieldNamespace, subfieldSchemas)
                mergedSubfieldSchemas.add(0, NULL_SCHEMA)
                subfieldBuilder.type(Schema.createUnion(mergedSubfieldSchemas)).withDefault(null)
            }
            mergedSchemas.add(assembler.endRecord())
        }
        return mergedSchemas
    }

    /**
     * Take in a Json field definition, and generate a nullable Avro field schema. For example:
     *
     * <pre>
     * {"type": ["number", { ... }]} -> ["null", "double", { ... }]
    </pre> *
     */
    private fun parseJsonField(
        fieldName: String,
        fieldNamespace: String?,
        fieldDefinition: JsonNode,
        appendExtraProps: Boolean,
        addStringToLogicalTypes: Boolean
    ): Schema {
        // Filter out null types, which will be added back in the end.
        val nonNullFieldTypes = getNonNullTypes(fieldName, fieldDefinition)
            .flatMap { fieldType: JsonSchemaType ->
                val singleFieldSchema = parseSingleType(
                    fieldName,
                    fieldNamespace,
                    fieldType,
                    fieldDefinition,
                    appendExtraProps,
                    addStringToLogicalTypes
                )
                if (singleFieldSchema.isUnion) {
                    return@flatMap singleFieldSchema.types.toList()
                } else {
                    return@flatMap listOf(singleFieldSchema)
                }
            }
            .distinct()
            .toMutableList()

        return if (nonNullFieldTypes.isEmpty()) {
            Schema.create(Schema.Type.NULL)
        } else {
            // Mark every field as nullable to prevent missing value exceptions from Avro / Parquet.
            if (!nonNullFieldTypes.contains(NULL_SCHEMA)) {
                nonNullFieldTypes.add(0, NULL_SCHEMA)
            }
            // Logical types are converted to a union of
            // logical type itself and string. The purpose is to
            // default the logical type field to a string,
            // if the value of the logical type field is invalid and
            // cannot be properly processed.
            if (nonNullFieldTypes
                    .stream().anyMatch { schema: Schema -> schema.logicalType != null } &&
                !nonNullFieldTypes.contains(STRING_SCHEMA) && addStringToLogicalTypes
            ) {
                nonNullFieldTypes.add(STRING_SCHEMA)
            }
            Schema.createUnion(nonNullFieldTypes)
        }
    }

    companion object {
        private const val TYPE = "type"
        private val NULL_SCHEMA = Schema.create(Schema.Type.NULL)
        private val STRING_SCHEMA = Schema.create(Schema.Type.STRING)
        private val LOGGER = LoggerFactory.getLogger(
            JsonToAvroSchemaConverter::class.java
        )

        fun getNonNullTypes(
            fieldName: String?,
            fieldDefinition: JsonNode
        ): List<JsonSchemaType> {
            return getTypes(fieldName, fieldDefinition)
                .filter { type: JsonSchemaType -> type != JsonSchemaType.NULL }
        }

        /**
         * When no type is specified, it will default to string.
         */
        private fun getTypes(fieldName: String?, fieldDefinition: JsonNode): List<JsonSchemaType> {
            val combinedRestriction = getCombinedRestriction(fieldDefinition)
            if (combinedRestriction.isPresent) {
                return listOf(JsonSchemaType.COMBINED)
            }
            val typeProperty = fieldDefinition[TYPE]
            if (typeProperty == null || typeProperty.isNull) {
                LOGGER.warn(
                    "Field \"{}\" has no type specification. It will default to string",
                    fieldName
                )
                return listOf(JsonSchemaType.STRING)
            }
            if (typeProperty.isArray) {
                return typeProperty.elements().asSequence()
                    .map { s: JsonNode -> JsonSchemaType.fromJsonSchemaTypeName(s.asText()) }
                    .toList()
            }
            if (typeProperty.isTextual) {
                return listOf(JsonSchemaType.fromJsonSchemaTypeName(typeProperty.asText()))
            }
            LOGGER.warn(
                "Field \"{}\" has unexpected type {}. It will default to string.",
                fieldName,
                typeProperty
            )
            return listOf(JsonSchemaType.STRING)
        }

        fun getCombinedRestriction(fieldDefinition: JsonNode): Optional<JsonNode> {
            if (fieldDefinition.has("anyOf")) {
                return Optional.of(
                    fieldDefinition["anyOf"]
                )
            }
            if (fieldDefinition.has("allOf")) {
                return Optional.of(
                    fieldDefinition["allOf"]
                )
            }
            return if (fieldDefinition.has("oneOf")) {
                Optional.of(fieldDefinition["oneOf"])
            } else Optional.empty()
        }

        private fun getStringSchema(fieldType: JsonSchemaType, fieldDefinition: JsonNode): Schema {
            val fieldSchema: Schema = if (fieldDefinition.has("format")) {
                when (fieldDefinition["format"].asText()) {
                    "date-time" -> {
                        LogicalTypes.timestampMicros()
                            .addToSchema(Schema.create(Schema.Type.LONG))
                    }
                    "date" -> {
                        LogicalTypes.date()
                            .addToSchema(Schema.create(Schema.Type.INT))
                    }
                    "time" -> {
                        LogicalTypes.timeMicros()
                            .addToSchema(Schema.create(Schema.Type.LONG))
                    }
                    else -> {
                        Schema.create(fieldType.avroType)
                    }
                }
            } else {
                Schema.create(fieldType.avroType)
            }
            return fieldSchema
        }
    }
}
